firmware-support/bittide-sys/src/scatter_unit.rs

```diff
@@ -1,82 +1,34 @@
+impl<const MEM_SIZE: usize> ScatterUnit<MEM_SIZE> {
+    /// Create a new [`ScatterUnit`] instance given a base address.
     ///
     /// # Safety
     ///
-    /// The FDT must be a valid description of the hardware components that this
-    /// code is running on.
-    pub unsafe fn from_fdt_node(node: &FdtNode) -> Result<Self, ComponentLoadError> {
-        let get_node = |path| {
-            node.children()
-                .find(|child| matches_fdt_name(child, path))
-                .ok_or(ComponentLoadError::FdtNodeNotFound(path))
-        };
-
-        let get_reg = |node: &fdt::node::FdtNode, component| {
-            node.reg()
-                .ok_or(ComponentLoadError::RegNotFound { component })?
-                .next()
-                .ok_or(ComponentLoadError::RegNotFound { component })
-        };
-
-        let memory_node = get_node("scatter-memory")?;
-        let metacycle_register_node = get_node("metacycle-reg")?;
-
-        let memory = get_reg(&memory_node, "memory")?;
-
-        if let Some(size) = memory.size {
-            if size != FRAME_SIZE {
-                return Err(ComponentLoadError::SizeMismatch {
-                    property: "memory frame size",
-                    expected: FRAME_SIZE,
-                    found: size,
-                });
-            }
+    /// The `base_addr` pointer MUST be a valid pointer that is backed
+    /// by a memory mapped scatter unit instance.
+    pub unsafe fn new(base_addr: *const ()) -> ScatterUnit<MEM_SIZE> {
+        let addr = base_addr as *const u8;
+        ScatterUnit {
+            memory: addr,
+            metacycle_register: addr.add(MEM_SIZE * 4),
         }
-
-        let metacycle_register = {
-            let reg = get_reg(&metacycle_register_node, "metacycle_register")?;
-            if let Some(size) = reg.size {
-                if size != 1 {
-                    return Err(ComponentLoadError::SizeMismatch {
-                        property: "metacycle register size",
-                        expected: 1,
-                        found: size,
-                    });
-                }
-            }
-            reg.starting_address
-        };
-
-        Ok(ScatterUnit {
-            memory: memory.starting_address,
-            metacycle_register,
-        })
-    }
-
-    pub const fn frame_size(&self) -> usize {
-        FRAME_SIZE
     }
 
-    pub fn read_frame_memory(&self, data: &mut [u8; FRAME_SIZE]) {
-        for (i, d) in data.iter_mut().enumerate() {
-            unsafe {
-                *d = self.memory.add(i).read_volatile();
-            }
-        }
+    /// # Safety
+    ///
+    /// The destination memory size must be smaller or equal to the size of the
+    /// `ScatterUnit`.
+    pub unsafe fn copy_to_slice(&self, dst: &mut [u8], offset: usize) {
+        let src = self.memory.add(offset);
+        assert!(dst.len() + offset <= MEM_SIZE * 4);
+        core::ptr::copy_nonoverlapping(src, dst.as_mut_ptr(), dst.len());
```

Comment:
Can be removed if `Deref` and `DerefMut` are implemented.

Reply:
Ignore, see comment on `copy_from_slice`, but replace the `write_volatile` with a `read_volatile` and assign that to the appropriate index in `dst`.